src/extension/control-flow/src/main/java/org/geoserver/flow/ControlFlowCallback.java

```diff
@@ -187,51 +188,43 @@ public Operation operationDispatched(Request request, Operation operation) {
 
     public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
         this.applicationContext = applicationContext;
-    }
-
-    @Override
-    public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory) throws BeansException {
-        if (GeoServerExtensions.bean(FlowControllerProvider.class) == null) {
-            beanFactory.initializeBean(provider, "defaultFlowControllerProvider");
+        if (applicationContext instanceof ConfigurableApplicationContext) {
+            // register default beans if needed
+            registDefaultBeansIfNeeded((ConfigurableApplicationContext) applicationContext);
+        } else {
+            // we cannot regist default beans, there is nothing else we can do about this
+            LOGGER.warning("Application context not configurable, control-flow default beans will not be registered.");
         }
-        
-        // look for a ControlFlowConfigurator in the application context, if none is found, use the
-        // default one
         provider = GeoServerExtensions.bean(FlowControllerProvider.class, applicationContext);
+        // default beans may have not been registered
         if (provider == null) {
             provider = new DefaultFlowControllerProvider(applicationContext);
         }
     }
 
-    @Override
-    public void postProcessBeanDefinitionRegistry(BeanDefinitionRegistry registry)
-            throws BeansException {
-        // Inject DefaultFlowControllerProvider definition it not defined
-        if (GeoServerExtensions.bean(FlowControllerProvider.class, applicationContext) == null) {
```

Comment:
Skipping GeoServerExtensions.bean breaks the ability to use ExtensionFilter to disable a certain plugin in favor of another. Not that bad here, only one FlowControllerProvider is meant to be used in the whole application regardless, so I'm +0 on this change.